Miki.Framework/MikiAppBuilder.cs

```diff
@@ -3,37 +3,52 @@
 
 namespace Miki.Framework
 {
-    public class MikiAppBuilder
-    {
-        public IServiceCollection Services { get; }
+	public class MikiAppBuilder
+	{
+		public IServiceCollection Services { get; }
 
-        public MikiAppBuilder()
-        {
-            Services = new ServiceCollection();
-        }
+		public MikiAppBuilder()
+		{
+			Services = new ServiceCollection();
+		}
 
-        public MikiAppBuilder AddSingletonService<T>(Func<IServiceProvider, T> factory)
+		public MikiAppBuilder AddSingletonService<T>(Func<IServiceProvider, T> factory)
+			where T : class
+		{
+			Services.AddSingleton(factory);
+			return this;
+		}
+        public MikiAppBuilder AddSingletonService<T>()
             where T : class
         {
-            Services.AddSingleton(factory);
+            Services.AddSingleton<T>();
             return this;
         }
-        public MikiAppBuilder AddSingletonService<T>(T value)
+        public MikiAppBuilder AddSingletonService<T, TImpl>()
+            where T : class
+            where TImpl : class, T
         {
-            Services.AddSingleton(typeof(T), value);
+            Services.AddSingleton<T, TImpl>();
             return this;
         }
-
+		public MikiAppBuilder AddSingletonService<T>(T value)
+		{
+			Services.AddSingleton(typeof(T), value);
+			return this;
+		}
         public MikiAppBuilder AddSingletonService(Type t, object value)
-        {
-            Services.AddSingleton(t, value);
-            return this;
-        }
+		{
+			Services.AddSingleton(t, value);
+			return this;
+		}
 
-        public MikiApp Build()
-        {
-            var services = Services.BuildServiceProvider();
-            return new MikiApp(services);
-        }
-    }
+		public MikiApp Build()
+		{
+			var app = new MikiApp();
+			Services.AddSingleton(app);
+			var services = Services.BuildServiceProvider();
+			app.Services = services;
+			return app;
```

Comment:
If we're registering MikiApp, couldn't we do the following?

```csharp
Services.AddSingleton<MikiApp>();

var provider = Services.BuildServiceProvider();
return provider.GetRequiredService<MikiApp>();
```

In my opinion is this a little bit cleaner since `IServiceProvider` gets automatically injected into MikiApp by `Microsoft.Extensions.DependencyInjection`.